File: BasicML/basicml/optim/nesterov.py
from .optimizer             import Optimizer
from typing                 import Optional
from basicml.regularization import Regularizer
from basicml.tensor         import Tensor
import numpy as np

class Nesterov(Optimizer):
    def __init__(self,
                 parameters : list[Tensor],
                 lr         : float,
                 regularizer: Optional[Regularizer] = None,
                 momentum   : float                 = 0.9):
        super().__init__(parameters, lr, regularizer)
        self.momentum   = momentum
        self.velocities = [np.zeros_like(p.data) for p in self.parameters]

    def step(self):
        for i, param in enumerate(self.parameters):
            if param.requires_grad and param.grad is not None:
                grad = param.grad
                if self.regularizer is not None:
                    grad = grad + self.regularizer.grad(param)
                previous            = self.velocities[i]
                self.velocities[i]  = self.momentum * previous - self.lr * grad
                param.data += -self.momentum * previous + (1.0 + self.momentum) * self.velocities[i]

    def zero_grad(self):
        for param in self.parameters:
            param.zero_grad()

File: AdvanceML/src/optim/nesterov.cpp
#include "advanceml/optim/nesterov.hpp"

namespace advanceml {

Nesterov::Nesterov(std::vector<Tensor> parameters, float learning_rate, float momentum)
    : parameters_(std::move(parameters)), learning_rate_(learning_rate), momentum_(momentum) {
    velocities_.reserve(parameters_.size());
    for (const Tensor& param : parameters_) {
        velocities_.emplace_back(param.data().size(), 0.0f);
    }
}

void Nesterov::step() {
    for (size_t i = 0; i < parameters_.size(); ++i) {
        Tensor& param = parameters_[i];
        if (!param.has_grad()) {
            continue;
        }
        Tensor grad = param.grad();
        std::vector<float>& values = param.mutable_data();
        std::vector<float>& velocity = velocities_[i];
        for (size_t j = 0; j < values.size(); ++j) {
            const float previous = velocity[j];
            velocity[j] = momentum_ * previous - learning_rate_ * grad.data()[j];
            values[j] += -momentum_ * previous + (1.0f + momentum_) * velocity[j];
        }
    }
}

void Nesterov::zero_grad() {
    for (Tensor& param : parameters_) {
        param.zero_grad();
    }
}

}  // namespace advanceml
